### PR TITLE
Move magnifier icon, add custom marker, add keyboard shortcuts

### PR DESCRIPTION
**Pull request by Pierre-Lefevre**

### CODE REVIEW

File: demo/index.html [hunk 1]
```diff
@@ -1,84 +1,178 @@
...
+      map.addControl(new PeliasGeocoder({
+        params: {'access-token': 'P4QzOCt4wWfPKfuPuMF1UQSupLKLsO7ZC4DBRF6GAoC0AcaSAKRvD4v2vix948q7'},
+        flyTo: 'hybrid',
+        wof: true,
+        url: 'https://places.jawg.io/v1',
+        useFocusPoint: true,
+        // marker: false,
+        // marker: true,
+        // marker: {
+        //   icon: 'beer-11',
```
Comment (Joxit) [inline demo/index.html]: What is that ?

File: demo/index.html [hunk 2]
```diff
@@ -1,84 +1,178 @@
...
+        height: 40px;
+        width: 40px;
+        position: absolute;
+        margin: -13px 0 0 -13px;
+        animation: pelias-ctrl-pulsate 1s ease-out;
+        animation-iteration-count: infinite;
+        opacity: 0;
+        box-shadow: 0 0 1px 2px #2999fd;
+        animation-delay: 1.1s;
+      }
```
Comment (Joxit) [inline demo/index.html:95]: Is this really used in the demo ?

File: pelias-geocoder.js [hunk 3]
```diff
@@ -1,109 +1,403 @@
...
+
+  this.opts = {
+    placeholder: opts.placeholder || 'Search',
+    url: opts.url,
+    flyTo: opts.flyTo === undefined ? true : opts.flyTo,
+    sources: opts.sources,
+    useFocusPoint: opts.useFocusPoint,
+    removeDuplicates: opts.removeDuplicates === undefined ? true : opts.removeDuplicates,
+    onSubmitOnly: opts.onSubmitOnly
+  };
```
Comment (Joxit) [inline pelias-geocoder.js]: I do not like this way of instantiation. I prefer the old version, sry :confused:

**Review by Joxit — CHANGES_REQUESTED** [review]

File: pelias-geocoder.js [hunk 4]
```diff
@@ -8,294 +13,429 @@ function PeliasGeocoder(opts) {
...
     this.opts.wof.fillOutlineColor = opts.wof.fillOutlineColor || "rgba(200, 40, 32, 0.7)";
+    this.getWOFURL = opts.wof.getWOFURL || this.getDefaultWOFURLFunction();
   }
+
   if (opts.params) {
     this.params = '';
-    for (var i in opts.params) {
-      this.params += '&' + i + '=' + opts.params[i];
+    for (var key in opts.params) {
+      if (opts.params.hasOwnProperty(key)) {
```
Comment (Joxit) [inline pelias-geocoder.js:36]: Whye use hasOwnProperty in a for each keys ? This should be always true right ?

File: pelias-geocoder.js [hunk 5]
```diff
@@ -8,294 +13,429 @@ function PeliasGeocoder(opts) {
...
-  el.className = 'pelias-ctrl-geocoder mapboxgl-ctrl';
-
-  var icon = document.createElement('span');
-  icon.className = 'geocoder-icon geocoder-icon-search';
+  this._buildWrapper();
+  this._buildIconSearch();
+  this._buildInputWrapper();
+  this._buildInputActionsWrapper();
+  this._buildInput();
+  this._buildResults();
```
Comment (Joxit) [inline pelias-geocoder.js]: Two many function call... All the stuff is hidden and harder to read what is really done. (Especially for the creation of HTML elements)

File: pelias-geocoder.js [hunk 6]
```diff
@@ -8,294 +13,429 @@ function PeliasGeocoder(opts) {
...
-}
+  return this._wrapperEl;
+};
 
-PeliasGeocoder.prototype.search = function(opts, callback) {
-  var self = this;
+PeliasGeocoder.prototype.search = function (opts, callback) {
   opts = opts || {};
-  if (!opts.text || opts.text.length == 0) {
+  if (!opts.text || opts.text.length === 0) {
```
Comment (Joxit) [inline pelias-geocoder.js:80]: :+1:

File: pelias-geocoder.js [hunk 7]
```diff
@@ -8,294 +13,429 @@ function PeliasGeocoder(opts) {
...
-  var self = this;
-  if (!self.opts.removeDuplicates) {
+  if (!this.opts.removeDuplicates) {
     return features;
   }
-  features.forEach(function(e) {
-    var label = e.properties.label;
+
+  var self = this;
+  features.forEach(function (feature) {
```
Comment (Joxit) [inline pelias-geocoder.js:129]: Good naming `feature` instead of `e` :smile:

File: pelias-geocoder.js [hunk 8]
```diff
@@ -8,294 +13,429 @@ function PeliasGeocoder(opts) {
...
-PeliasGeocoder.prototype._areNear = function(c1, c2, tolerance) {
+PeliasGeocoder.prototype._areNear = function (c1, c2, tolerance) {
   return this._between(c1[0], c2[0] - tolerance, c2[0] + tolerance) && this._between(c1[1], c2[1] - tolerance, c2[1] + tolerance);
-}
+};
 
-PeliasGeocoder.prototype._latlngToArray = function(center) {
-  return [center.lng, center.lat];
-}
+PeliasGeocoder.prototype._coordinatesToArray = function (coordinates) {
```
Comment (Joxit) [inline pelias-geocoder.js:166]: Cool naming :wink:

File: pelias-geocoder.js [hunk 9]
```diff
@@ -8,294 +13,429 @@ function PeliasGeocoder(opts) {
...
+    for (var i = 0; i < this._customHtmlMarkers.length; ++i) {
+      this._customHtmlMarkers[i].remove();
+    }
+    this._customHtmlMarkers = [];
+  } else {
+    this._removeSources(this.opts.marker, this.markerLayerId)
+  }
+};
+
+PeliasGeocoder.prototype._buildWrapper = function () {
```
Comment (Joxit) [inline pelias-geocoder.js]: Overkill function

File: pelias-geocoder.js [hunk 10]
```diff
@@ -8,294 +13,429 @@ function PeliasGeocoder(opts) {
...
+  this._iconSearchEl.addEventListener("click", function () {
+    if (self._results && self._results.features[0]) {
+      var feature = self._results.features[0];
+      self._goToFeatureLocation(feature);
+      self._updateMarkers(feature);
+    }
+  });
+};
+
+PeliasGeocoder.prototype._buildInputWrapper = function () {
```
Comment (Joxit) [inline pelias-geocoder.js]: Overkill function

File: pelias-geocoder.js [hunk 11]
```diff
@@ -8,294 +13,429 @@ function PeliasGeocoder(opts) {
...
+  this._inputWrapperEl = document.createElement('div');
+  this._inputWrapperEl.className = 'input-wrapper';
+};
+
+PeliasGeocoder.prototype._buildInputActionsWrapper = function () {
```
Comment (Joxit) [inline pelias-geocoder.js]: Overkill function

File: pelias-geocoder.js [hunk 12]
```diff
@@ -8,294 +13,429 @@ function PeliasGeocoder(opts) {
...
+    }
+    if (e.keyCode === self._keyCodes.arrowDown && self._resultsEl.childNodes[index + 1]) {
+      self._resultsEl.childNodes[index + 1].focus();
+    }
+  });
+
+  return resultEl;
+};
+
+PeliasGeocoder.prototype._sendXmlHttpRequest = function (url, callback) {
```
Comment (Joxit) [inline pelias-geocoder.js:408]: That should be usefull!

**Review by Joxit — CHANGES_REQUESTED** [review]

File: pelias-geocoder.js (hunk 4 above)
Comment (Pierre-Lefevre) [reply pelias-geocoder.js:36]: It allows to only consider the properties attached to the object itself, and not all properties on the prototype chain (that can be added by someone else).

Comment (Joxit) [reply pelias-geocoder.js:36]: :+1:

**Review by Joxit — APPROVED** [review]
Sounds good :smile:

**Pull request merged**